### ISSUE TITLE
Split up param.yaml by group and rename groups

### ISSUE THREAD

**Issue by hmgaudecker** [iza-institute-of-labor-economics/gettsim#79]
### Current and desired situation

param.yaml is a monster and unreadable. We have already agreed to split it up in smaller files. This makes it concrete and suggests renaming groups to something comprehensible by the non-initiated.

### Proposed implementation

Use something close to the current groups (https://github.com/iza-institute-of-labor-economics/gettsim/pull/56#issue-335872991) with more sensible and more readable names. This table is ordered by proposed name so that it is clear which ones I propose to merge:

| Current           | Proposal        |
| ----------------- | --------------- |
| abgeltung         | abg_st          |
| alg               | arb_los_geld    |
| alg2              | arb_los_geld_2  |
| alh               | arb_los_hilfe   |
| kinderfreibetrag  | ek_st_abzuege   |
| zve               | ek_st_abzuege   |
| tarif             | ek_st_tarif     |
| tax_sched         | ek_st_tarif     |
| eg                | erziehungsgeld  |
| kindergeld        | kindergeld      |
| kiz               | kinderzuschlag  |
| ust               | mw_st           |
| rente             | rente           |
| soc_ins_contrib   | soz_vers_beitr  |
| sgbxii            | sozialhilfe     |
| sh                | sozialhilfe     |
| uhv               | unterh_vorsch   |
| wohngeld          | wohngeld        |

Please check carefully before we go for this (@mjbloemer, @Eric-Sommer, @stichnoth, your opinions woud be especially great to have!) - I may have missed something or misunderstood what is actually meant by a parameter.

Does it make sense to include *abgeltungssteuer* as a parametr of *ek_st_tarif*? It seems weird to have a group with just one element.

**Comment by mjbloemer** [comment]
I think coarse groups are good. A few quick comments:

The names are really a matter of taste, so in that fashion:
- I don't like the *_st* abbreviations *ek_st*, *mw_st* (I prefer *est* and *ust* or *vat*).
- *arb_los_geld* sounds strange.
- Probably *sozialhilfe* is not a good category for the all the *sgbxii* parameters.
- *_abzuege* is good and more precise than zve.
- I would drop the *_beitr* in *soz_vers_beitr*. It is just a group with everything SSC related.

> Does it make sense to include abgeltungssteuer as a parametr of ek_st_tarif? It seems weird to have a group with just one element.

We might put Abgeltungssteuer in an ESt group. But it is simply not a parameter of the income tax *tarif*. People will be really confused if we did that. Same for the Soli.

**Comment by hmgaudecker** [comment]
> I don't like the __st_ abbreviations _ek_st_, _mw_st_ (I prefer _est_ and _ust_ or _vat_).

I think this depends on whether you are familiar with the laws' abbreviations or not. But fair enough. I would like to keep the underscores, though (more Pythonic :-))

>   _arb_los_geld_ sounds strange.

Why? Seems natural to me... Of course, if you are a German native speaker and familiar with the debate, _alg_ is fine. 

>  Probably _sozialhilfe_ is not a good category for the all the _sgbxii_ parameters.

But for the ones we have in there currently it would be, no? I count 7 parameters in that group, 5 of which are _Sozialhilfe Mehrbedarfszuschläge_, the other two are mysterious to me. If we go deeply into Grundsicherung im Alter / bei Pflegebedürftigkeit / ..." that might merit another group. 

>  I would drop the __beitr_ in _soz_vers_beitr_. It is just a group with everything SSC related.

"SSC" = social security contributions? Precisely, hence the "beitr" :-). For the payouts, things like "ges_renten_vers" or "arb_los_geld" are separate. I would like to make that clear in the grouping.

> > Does it make sense to include abgeltungssteuer as a parametr of ek_st_tarif? It seems weird to have a group with just one element.
> 
> We might put Abgeltungssteuer in an ESt group. But it is simply not a parameter of the income tax _tarif_. People will be really confused if we did that. Same for the Soli.

Fair enough. 

Updated proposal:

| Current           | Proposal        |
| ----------------- | --------------- |
| abgeltung         | abg_st          |
| alg               | arb_los_geld    |
| alg2              | arb_los_geld_2  |
| alh               | arb_los_hilfe   |
| kinderfreibetrag  | e_st_abzuege    |
| zve               | e_st_abzuege    |
| tarif             | e_st_tarif      |
| tax_sched         | soli_st_tarif   |
| eg                | erziehungsgeld  |
| kindergeld        | kindergeld      |
| kiz               | kinderzuschlag  |
| ust               | u_st            |
| rente             | ges_renten_vers |
| soc_ins_contrib   | soz_vers_beitr  |
| sgbxii            | sozialhilfe     |
| sh                | sozialhilfe     |
| uhv               | unterh_vorsch   |
| wohngeld          | wohngeld        |

**Comment by mjbloemer** [comment]
> > _arb_los_geld_ sounds strange.
> 
> Why? Seems natural to me... Of course, if you are a German native speaker and familiar with the debate, _alg_ is fine.

Should the underscores signal the original words of the compound Arbeitslosengeld? Why?

If so: Arbeitslosengeld is a nominal composition of Arbeitslos and Geld. But the *los* in Arbeitslos is simply a negating suffix used for the formation of the one noun. At least the underscore `_los` seems wrong.

For me as a native speaker this sounds strange an I don't think it helps non native speakers. If you are not a native speaker you would probably end looking up *arb*, *los* and *geld* in a dictionary. Will you find something for *arb*? What will you get for *los*?

**Comment by mjbloemer** [comment]
And: When we use *arb_los_geld* why not *kinder_geld*, *wohn_geld*, *kinder_zuschlag*, *erziehungs_geld*, *sozial_hilfe*...? It also has a bit of Bild-Zeitungs-Bindestrich ;)

**Comment by hmgaudecker** [comment]
Haha, I like strong opinions! 

I used underscores to signal abbreviations, more than in a precise syntactical sense. The one exception is _ges_renten_vers_, but I would be open to _ges_rentenvers_ there, too. 

Would be great to have some more opinions here! @Eric-Sommer @stichnoth ?

**Comment by Eric-Sommer** [comment]
- The two `sozhilfe` items refer to two different regimes. Before 2004, social assistance was an important pillar of the benefit system. The current SGB XII stuff basically fills some loopholes which are left by ALG2 etc. So I'd change the 2nd one `sozhilfe_pre2004` or similar.
- `abg_st`: `spsparf` and `spwerbz` kind of also belong to that regime, but only since 2009. And they are not for Abgeltungssteuer exclusively (#81). If you don't like a group with one member, you can as well put it to other `ESt` stuff.
- If `abg_st` is an own group, why is `soli` not? It's a different tax with an own law. Btw, the 0.2 share in `soli_formula` should also be in param.
- `eg` is Elterngeld since 2008 (?)
- I find `arb_los` weird, too. why not `alo_geld`, `alo_geld2`, `alo_hilfe`?
- `e_st` and `u_st` take the legal abbrevation and add an underscore, which does not really clear things up IMO. I'd prefer `est` and `mwst`. The whole world calls the thing VAT/Mehrwertsteuer except for German and Austrian tax lawyers.

**Comment by hmgaudecker** [comment]
> The two `sozhilfe` items refer to two different regimes. Before 2004, social assistance was an important pillar of the benefit system. The current SGB XII stuff basically fills some loopholes which are left by ALG2 etc. So I'd change the 2nd one `sozhilfe_pre2004` or similar.

I think these things are better left for implementation details inside one group. Let us please not put any years in the group names. 

> `abg_st`: `spsparf` and `spwerbz` kind of also belong to that regime, but only since 2009. And they are not for Abgeltungssteuer exclusively (#81). If you don't like a group with one member, you can as well put it to other `ESt` stuff.

Will have to check these two in more detail. We will have some parameters where it is not fully clear in which group to put them, but that is the nature of things. 

> If `abg_st` is an own group, why is `soli` not? It's a different tax with an own law. 

It is  in the revised version :-). @mjbloemer already convinced me on that one. 

> Btw, the 0.2 share in `soli_formula` should also be in param.

Feel free to open an issue! 

> * `eg` is Elterngeld since 2008 (?)

There was no Elterngeld before, right? Question is whether we want a more generic name then and include both regimes or have two categories? 

Kinda goes back to the first thing: the old Sozialhilfe was split up, but parts of it still has the same name... To me it is clearest to go by (sets of) laws (but with clear names, not _sgbxii_ etc.). 

> I find `arb_los` weird, too. why not `alo_geld`, `alo_geld2`, `alo_hilfe`?

Cool, even more suggestions :-). To me, it is just obvious how I would fill in "arb_los", this is not the case for "alo" (or "alg" if I am not familiar wih the debate).

This said, I guess I could live with "alg" if all of you insist. 

> `e_st` and `u_st` take the legal abbrevation and add an underscore, which does not really clear things up IMO.

No, it takes the legal CamelCase-style _ESt_ and converts it to lowercase_with_undercores, which is   Python style... 

> I'd prefer `est` and `mwst`. The whole world calls the thing VAT/Mehrwertsteuer except for German and Austrian tax lawyers.

That is for @mjbloemer and you to clear up, I do not have a strong opinion on u / mw.

**Comment by Eric-Sommer** [comment]
> I think these things are better left for implementation details inside one group. Let us please not put any years in the group names.
OK
> > If `abg_st` is an own group, why is `soli` not? It's a different tax with an own law.
> 
> It is in the revised version :-). @mjbloemer already convinced me on that one.

I only see `soli_st_tarif` encompassing tax schedule and soli. 

> There was no Elterngeld before, right? Question is whether we want a more generic name then and include both regimes or have two categories?
If we can stand one group for sozialhilfe, we can do it here also. Elterngeld stuff starts with `elg`, Erzierhungsgeld with `eg`. 
 
> This said, I guess I could live with "alg" if all of you insist.

and what about Arbeitslosenhilfe? `alh`?

**Comment by hmgaudecker** [comment]
> > If `abg_st` is an own group, why is `soli` not? It's a different tax with an own law.
> > It is in the revised version :-). @mjbloemer already convinced me on that one.
> 
> I only see `soli_st_tarif` encompassing tax schedule and soli.

Well, the group _tax_sched_ currently has two entries: "Soli-Freigrenze" and "Höhe des Solidaritätszuschlags auf die Einkommensteuer" ... 

> > There was no Elterngeld before, right? Question is whether we want a more generic name then and include both regimes or have two categories?
> 
> If we can stand one group for sozialhilfe, we can do it here also. Elterngeld stuff starts with `elg`, Erzierhungsgeld with `eg`.

The cases are slightly different because current Sozialhilfe is still called that way, but only a subset of the previous regime. Here the names are different. I am fine with either -- I would definitely look up the pre-2008 parameters also under the name "Elterngeld" -- but it would be good to have more opinions.

> > This said, I guess I could live with "alg" if all of you insist.
> 
> and what about Arbeitslosenhilfe? `alh`?

Sure. That alh and alg are more easily mixed up when scanning through the code is one reason for a more verbose naming from my point of view.

**Comment by hmgaudecker** [comment]
Seems like discussion has petered out a little here. Let us please have some last round of votes from everyone. My current proposal:

| Current           | Proposal        |
| ----------------- | --------------- |
| abgeltung         | abgelt_st       |
| alg               | arb_los_geld    |
| alg2              | arb_los_geld_2  |
| alh               | arb_los_hilfe   |
| kinderfreibetrag  | e_st_abzuege    |
| zve               | e_st_abzuege    |
| tarif             | e_st_tarif      |
| tax_sched         | soli_st_tarif   |
| eg                | erziehungsgeld  |
| kindergeld        | kindergeld      |
| kiz               | kinderzuschlag  |
| ust               | u_st            |
| rente             | ges_renten_vers |
| soc_ins_contrib   | soz_vers_beitr  |
| sgbxii            | sozialhilfe     |
| sh                | sozialhilfe     |
| uhv               | unterh_vorsch   |
| wohngeld          | wohngeld        |

Controversial are, if you do not want to go through the whole discussion above:

- `arb_los_geld` / `arb_los_hilfe` vs. `alg` / `alh`
- `erziehungsgeld` / `elterngeld` vs. merging both in `elterngeld`  
- `u_st` vs. `mw_st`

This was a quick summary, if I missed anything close to your hearts, @Eric-Sommer & @mjbloemer, please chime in.

Please cast your votes and if required, add (further) comments, @ChristianZimpelmann,@Eric-Sommer,@janosg,@MaxBlesch,@mjbloemer,@stichnoth,@tobiasraabe!

**Comment by mjbloemer** [comment]
+1 for the list

Except the controversial cases:
* +1 `alg`/`alg2`/`alh` (or a -1 on `arb_los_*`)
* +1 on `erziehungsgeld` / `elterngeld`
* 0 vs. 0 on `u_st` vs. `mw_st`

missed points (further discussion is optional):
* why is `soli_st_tarif` considered a tarif and `abgelt_st` not (it is a flat tarif). Can't we drop the `_tarif` altogether?
* were there considerations to align the group names with the names of the .py files? Why have a kinderzuschlag.yaml and a benefits/kiz.py or a ek_st_abzuege.yaml and a taxes/zve.py?

**Comment by hmgaudecker** [comment]
> why is `soli_st_tarif` considered a tarif and `abgelt_st` not (it is a flat tarif). 

Path dependence :-) Good catch!

> Can't we drop the `_tarif` altogether?

I think that is uncontroversial, I edited the list correspondingly!

> were there considerations to align the group names with the names of the .py files? Why have a kinderzuschlag.yaml and a benefits/kiz.py or a ek_st_abzuege.yaml and a taxes/zve.py? Similar for soz_vers_beitr.yaml.

I have not looked at that at all; that would of course make a lot of sense.

**Comment by MaxBlesch** [comment]
First I would align everything once we agreed on names!

My votes are:
+1 `m_st`
+1 `erziehungsgeld` / `elterngeld`

The other debate is more difficult. I am definitely -1 `alg` / `alg2` / `alh`, as I think this too short and will be difficult for non German speakers. For me the _ between "arb" and "los" is a bit weird too. But I still think that is the best compromise/way to do it so +1 for that.

**Comment by ChristianZimpelmann** [comment]
I also think `arb_los_geld` looks a bit weird.
How about `arbeitsl_geld, arbeitsl_geld2, arbeitsl_hilfe`?

+1 `mw_st`
+1 merging both in `elterngeld` (It would be even better if we had a category name that includes both elterngeld and erziehungsgeld)
+1 for the rest of the list

**Comment by hmgaudecker** [comment]
I like that suggestion on the unemployed.

If I counted correctly, including my own preferences and assuming transitivity on @MaxBlesch's preferences:

* 3 / 1 on verbose `arbeitsl_x` vs. short `alx`
* 2 / 1 on separate `elterngeld` / `erziehungsgeld` vs. merged `elterngeld` (might change if somebody had a good category name as suggested by @ChristianZimpelmann, but I could not come up with any)
* 2 / 0 on `mw_st` vs `u_st`

Updated list given current votes:

| Current           | Proposal        |
| ----------------- | --------------- |
| abgeltung         | abgelt_st       |
| alg               | arbeitsl_geld   |
| alg2              | arbeitsl_geld_2 |
| alh               | arbeitsl_hilfe  |
| kinderfreibetrag  | e_st_abzuege    |
| zve               | e_st_abzuege    |
| tarif             | e_st            |
| tax_sched         | soli_st         |
| eg                | erziehungsgeld  |
| elg               | elterngeld      |
| kindergeld        | kindergeld      |
| kiz               | kinderzuschlag  |
| ust               | mw_st           |
| rente             | ges_renten_vers |
| soc_ins_contrib   | soz_vers_beitr  |
| sgbxii            | sozialhilfe     |
| sh                | sozialhilfe     |
| uhv               | unterh_vorsch   |
| wohngeld          | wohngeld        |

I suggest keeping this open until Sunday, if nothing changes, @MaxBlesch can go ahead with the implementation next week.

**Comment by mjbloemer** [comment]
Once the split up is done, we should also remove the group key in the yaml files.

**Comment by hmgaudecker** [comment]
Alright, given there have been no further complaints, @MaxBlesch please go ahead implementing this. 

Please do two PRs, so we can track things later more easily.

1. Pure renaming in params.yaml
2. Split up, remove groups, pass around the 17 categories above to the functions instead of the `tb` monster.

Thanks in advance!
